substrate/primitives/staking/src/delegation.rs

```diff
@@ -0,0 +1,160 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use codec::{FullCodec, MaxEncodedLen};
+use scale_info::TypeInfo;
+use sp_runtime::{DispatchResult, Saturating};
+use sp_std::ops::Sub;
+
+/// Allows an account to accept stake delegations and manage its operations.
+pub trait DelegationInterface {
+	/// Balance type used by the staking system.
+	type Balance: Sub<Output = Self::Balance>
+		+ Ord
+		+ PartialEq
+		+ Default
+		+ Copy
+		+ MaxEncodedLen
+		+ FullCodec
+		+ TypeInfo
+		+ Saturating;
+
+	/// AccountId type used by the staking system.
+	type AccountId: Clone + sp_std::fmt::Debug;
+
+	/// Total delegated balance to this account.
+	fn delegated_balance(who: &Self::AccountId) -> Self::Balance;
+
+	/// Total delegated balance to this account that is not yet bonded to staking.
+	fn unbonded_balance(who: &Self::AccountId) -> Self::Balance;
+
+	/// Set intention to accept delegations.
+	fn accept_delegations(
+		delegatee: &Self::AccountId,
+		reward_destination: &Self::AccountId,
+	) -> DispatchResult;
+
+	/// Migrate an nominator account into a delegatee.
+	///
+	/// # Arguments
+	///
+	/// * `new_delegatee`: This is the current nominator account. Funds will be moved from this
+	///   account to `proxy_delegator` and delegated back to `new_delegatee`.
+	/// * `proxy_delegator`: All existing staked funds will be moved to this account. Future
+	///   migration of funds from `proxy_delegator` to `delegator` is possible via calling
+	///   [`Self::migrate_delegator`].
+	///  * `payee`: Delegatees need to set where they want their rewards to be paid out.
+	///
+	/// This is similar to [`Self::accept_delegations`] but allows a current nominator to migrate to
+	/// a delegatee.
+	fn migrate_accept_delegations(
```

Comment:
Yeah good points. I could remove it from the interface and have only this locally to be used in migration script. Though the other migrate function (for migrating delegators) is long term as np delegators will slowly migrate their funds.

Reply:
> Though the other migrate function (for migrating delegators) is long term as np delegators will slowly migrate their funds.

Makes sense. Maybe a separate interface for the passive migration would be nice in case other projects don't need to migrate, or in the case of the staking parachain were we could start with a fully migrated storage. In that case, its easy to remove the migration related code without affecting the main trait.